cia.mk

```
@@ -1,59 +1,91 @@
+	rm -rf ${vc_rom_dirs} ${vc_dir}/seeddb.bin
+
 
-$(vc_rom_dir): $(vc_orig_cia) $(vc_dir)/seeddb.bin
+# Actual rules
+
+# TODO: relying on directory modification time is a bad idea!
+# Silence `ctrtool`, which is VERY verbose by default
+${vc_dir}/%/: ${vc_dir}/%.orig.cia ${vc_dir}/seeddb.bin
 	mkdir -p $@
-	ctrtool --contents=$@/contents $@.orig.cia
-	ctrtool --seeddb=$(vc_dir)/seeddb.bin \
-	        --exheader=$@/exheader.bin \
-	        --exefsdir=$@/exefs \
-	        --romfsdir=$@/romfs \
-	        --logo=$@/logo.lz \
-	        --plainrgn=$@/plain.bin \
-	        $@/contents.0000.*
-	ctrtool --seeddb=$(vc_dir)/seeddb.bin \
-	        --romfsdir=$@/manual \
-	        $@/contents.0001.*
-	rm -f $@/contents.*
-	rm -f $@/romfs/rom/*
-	rm -f $@/romfs/*.patch
-
-$(join $(addsuffix /romfs/, $(vc_rom_dir)), $(vc_patch)): $(vc_rom_dir) ./$(vc_patch)
-	cp ./$(@F) $@
-	
-$(join $(addsuffix /romfs/rom/, $(vc_rom_dir)), $(vc_gbc)): $(vc_rom_dir) ./$(vc_gbc)
-	cp ./$(@F) $(basename $@)
-
-$(vc_game_cxi): $(vc_rom_dir) $(join $(addsuffix /romfs/, $(vc_rom_dir)), $(vc_patch)) $(join $(addsuffix /romfs/rom/, $(vc_rom_dir)), $(vc_gbc))
-	(cd $(basename $(basename $@))/; \
-	    makerom -f cxi -o ../../$(basename $@).cxi -rsf ../game.rsf \
-	            -exheader exheader.bin \
-	            -logo logo.lz \
-	            -plainrgn plain.bin \
-	            -code exefs/code.bin \
-	            -icon exefs/icon.bin \
-	            -banner exefs/banner.bin \
-	)
-
-$(vc_manual_cfa): $(vc_dir)/manual.rsf
-	makerom -f cfa -o $@ -rsf $(vc_dir)/manual.rsf
-
-$(vc_cia): $(vc_game_cxi) $(vc_manual_cfa)
-	makerom -f cia -o $@ -content $(basename $@).game.cxi:0:0 -content $(basename $@).manual.cfa:1:1
\ No newline at end of file
+	ctrtool --contents=$@contents $< >/dev/null
+	ctrtool --seeddb=${vc_dir}/seeddb.bin \
+	        --exheader=$@exheader.bin \
+	        --exefsdir=$@exefs \
+	        --romfsdir=$@romfs \
+	        --logo=$@logo.lz \
+	        --plainrgn=$@plain.bin \
+	        $@contents.0000.* >/dev/null
+	ctrtool --seeddb=${vc_dir}/seeddb.bin \
+	        --romfsdir=$@manual \
+	        $@contents.0001.* >/dev/null
+	rm -f $@contents.*
+	rm -f $@romfs/rom/*
+	rm -f $@romfs/*.patch
+
+# romfs files have the pattern appear twice in the path, which breaks pattern rules; we have to use `eval` instead
+# Careful that the contents of the `define`s are expanded twice:
+# 1. In the `call` function, and
+# 2. In the `eval` function.
+
+define copy_patch_rule
+$${vc_dir}/$(1)/romfs/$(1).patch: $(1).patch | ${vc_dir}/$(1)/
+	mkdir -p $${@D}
+	cp -T $$< $$@
+endef
+$(foreach vc,${vc_name},$(eval $(call copy_patch_rule,${vc})))
+
+define copy_rom_rule
+$${vc_dir}/$(1)/romfs/rom/$(1): $(1).gbc | ${vc_dir}/$(1)/
+	mkdir -p $${@D}
+	cp -T $$< $$@
+endef
+$(foreach vc,${vc_name},$(eval $(call copy_rom_rule,${vc})))
+
+define make_cxi_rule
+$${vc_dir}/$(1).game.cxi: $${vc_dir}/game.rsf $${vc_dir}/$(1)/romfs/$(1).patch $${vc_dir}/$(1)/romfs/rom/$(1)
```

To go with the vc\_cxi\_req I suggested earlier.

```suggestion
$${vc_dir}/$(1).game.cxi: $${vc_dir}/game.rsf $${vc_dir}/$(1)/romfs/$(1).patch $${vc_dir}/$(1)/romfs/rom/$(1) $(addprefix $${vc_dir}/$(1)/, $(vc_cxi_req))
```

This command should be re-run when any of the `vc_cxi_req` files are updated. The only issue is i'm not sure how to re-run `${vc_dir}/%/:` if one of these files are deleted, or missing.

---

The fix is to simply make the "directory rule" have these files as targets instead.